flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1023,25 +1014,63 @@ protected AbstractStateBackend createInnerBackend(Configuration config) {
 	// ------------------------------------------------------------------------
 	// ------------------------------------------------------------------------
 
+	private enum Event {
+		TASK_IS_RUNNING,
+	}
+
 	private static class EmptyStreamTask extends StreamTask<String, AbstractStreamOperator<String>> {
 
-		public EmptyStreamTask(Environment env) {
+		private boolean isRunningFoo;
+		private final LinkedBlockingQueue<Event> eventQueue;
+		private final OperatorChain<String, AbstractStreamOperator<String>> overrideOperatorChain;
+		private volatile boolean sourceFinished;
+
+		EmptyStreamTask(Environment env, LinkedBlockingQueue<Event> eventQueue, OperatorChain<String, AbstractStreamOperator<String>> operatorChain) {
 			super(env, null);
+			this.eventQueue = eventQueue;
+			this.overrideOperatorChain = operatorChain;
 		}
 
 		@Override
-		protected void init() throws Exception {}
+		protected void init() {
+			if (overrideOperatorChain != null) {
+				super.operatorChain = this.overrideOperatorChain;
+				super.headOperator = super.operatorChain.getHeadOperator();
+			}
+		}
 
 		@Override
-		protected void performDefaultAction(ActionContext context) throws Exception {
-			context.allActionsCompleted();
+		protected void performDefaultAction(ActionContext context) {
```

Comment:
I'm not sure if overriding this method is a good way to perform the test checks. I think it might be better to rebase this on top of
https://github.com/apache/flink/pull/8826
and implement `sourceFinished` and `emitEvent(Event.TASK_IS_RUNNING);` as an `StreamInputProcessor` there.

Reply:
I have changed the way how the tests are waiting for task to be running. That way it should not conflict with the future changes in your PR.